### PR TITLE
🐛 `TrajectoryData`: Add `pbc`

### LINKED ISSUE #6376: TrajectoryData does not preserve periodicity information
How to turn a non-periodic structure to periodic structure in few simple steps, going from StructureData to TrajectoryData and back again

```python
In [92]: s = StructureData(cell=None, pbc=(False, False, False))

In [93]: s.append_atom(position=[0.0, 0.0, 0.0], symbols='H')

In [94]: t = TrajectoryData(structurelist=(s,))

In [95]: t.get_step_structure(0).pbc
Out[95]: (True, True, True)
```

trying to store such a transformed structure leads to sadness (predictably)

```python
----> 1 t.get_step_structure(0).store()

~/atmospec/aiida-core/src/aiida/orm/nodes/node.py in store(self)
    522             self._validate_storability()
--> 523             self._validate()
    524 
~/atmospec/aiida-core/src/aiida/orm/nodes/data/structure.py in _validate(self)
    920 
--> 921         _validate_dimensionality(self.pbc, self.cell)
    922 
~/atmospec/aiida-core/src/aiida/orm/nodes/data/structure.py in _validate_dimensionality(pbc, cell)
   2458     if dim['value'] == 0:
-> 2459         raise ValueError(f'Structure has periodicity {pbc} but {dim["dim"]}-d volume 0.')
```

Looking at the code, it seems like TrajectoryData does not track periodicity information at all, which is a bad news for everything except 3D-periodic structures.

This is currently a big issue for my app, since I am using TrajectoryData extensively to store molecular conformers. 

cc @mbercx

### LINKED ISSUE #7037: pbc in `TrajectoryData`
Dear all,
I would like to propose adding a pbc array to `TrajectoryData`.

Recently, in (https://github.com/aiidateam/aiida-quantumespresso/commit/20ac738c4bef22196ecf4a2c974ecb94cdd8e456), we introduced the possibility to use `TrajectoryData` as input for NEB calculations in AiiDA-quantumespresso.
In this context, having both the cell and the pbc configuration would allow for the correct generation of the k-points grid.
More generally, I believe that storing the pbc information is useful to preserve knowledge about the system’s dimensionality.

### PR DESCRIPTION
**Pull request by mbercx**
>[!NOTE]
> There is a lot of code debt in the `TrajectoryData`, and the API is quite old and not optimal. Fixing all of that is beyond the scope of this PR, and probably we should redesign it in `aiida-atomistic`.

Fixes https://github.com/aiidateam/aiida-core/issues/6376
Fixes https://github.com/aiidateam/aiida-core/issues/7037

Pinging @danielhollas, @rikigigi and @dbidoggia for review, since they raised the original issues.

The current changes try to remain backwards-compatible for `get_step_data`, i.e. it now keeps on returning the same tuple of length 5. However, this does mean that the method does not return the `pbc`. For the `get_step_structure` method, I directly set the `pbc` from the property obtained from the attributes.

### CODE REVIEW

File: src/aiida/orm/nodes/data/array/trajectory.py [hunk 1]
```diff
@@ -131,14 +137,20 @@ def set_trajectory(self, symbols, positions, stepids=None, cells=None, times=Non
...
+            length three with booleans indicating if the structure is periodic in that
+            direction. The same periodic boundary conditions are set for each step.
 
         .. todo :: Choose suitable units for velocities
         """
         import numpy
 
-        self._internal_validate(stepids, cells, symbols, positions, times, velocities)
-        # set symbols as attribute for easier querying
+        pbc = pbc or [True, True, True]
```
Comment (danielhollas) [inline src/aiida/orm/nodes/data/array/trajectory.py]:
```suggestion
        pbc = pbc or [True, True, True]
```
This default makes me uneasy. Should we instead look at the `cells` and see if it is defined and non-zero?

Comment (mbercx) [reply src/aiida/orm/nodes/data/array/trajectory.py]: Sensible check! My MC3D bias is showing again ^^

Comment (mbercx) [reply src/aiida/orm/nodes/data/array/trajectory.py]: > if it is defined and non-zero?

Can the `cells` be defined as zero? 🤔

Comment (dbidoggia) [reply src/aiida/orm/nodes/data/array/trajectory.py]: Do not know when I would use a structure without specifying cells, but agree to consider them `False` if not specified

Comment (mbercx) [reply src/aiida/orm/nodes/data/array/trajectory.py]: Apparently you can set e.g. `cell = [[0, 0, 0], [0, 1, 0], [0, 0, 1]]`, but I'm not sure that's sensible? @danielhollas right now I've adapted the default to be `[False, False, False]` in case `cells=None`. Is that sufficient? Or what did you have in mind?

File: src/aiida/orm/nodes/data/array/trajectory.py [hunk 2]
```diff
@@ -131,14 +137,21 @@ def set_trajectory(self, symbols, positions, stepids=None, cells=None, times=Non
...
+            length three with booleans indicating if the structure is periodic in that
+            direction. The same periodic boundary conditions are set for each step.
 
         .. todo :: Choose suitable units for velocities
         """
         import numpy
 
-        self._internal_validate(stepids, cells, symbols, positions, times, velocities)
-        # set symbols as attribute for easier querying
+        pbc_default = [True, True, True] if cells is not None else [False, False, False]
```
Comment (danielhollas) [inline src/aiida/orm/nodes/data/array/trajectory.py]: Perhaps we should print a deprecation warning here? This is in any case a best-guess since this doesn't support e.g. 2D structures. So I think we should deprecate the possibility of "not explicitly specifying PBC"  (with a possible exception of allowing for `None` if cell is `None`?)

**Review by danielhollas — CHANGES_REQUESTED** [review]
@mbercx thank you very much for working on this. I'd like to also try this out in my app to verify.

Also, can you please add test cases for the default PBC behaviour?

**Comment by mbercx** [comment]
@danielhollas if you have a moment, could you have another look? We could still squeeze this into 2.7.2, since it's fixing a bug in my view. I want to avoid having this open too long and it falling through the cracks. 🙏

File: src/aiida/orm/nodes/data/array/trajectory.py [hunk 3]
```diff
@@ -131,14 +141,28 @@ def set_trajectory(self, symbols, positions, stepids=None, cells=None, times=Non
...
         import numpy
 
-        self._internal_validate(stepids, cells, symbols, positions, times, velocities)
-        # set symbols as attribute for easier querying
+        if cells is None:
+            pbc = pbc or [False, False, False]
+        elif pbc is None:
+            warn(
+                "When 'cells' is not None, the periodic boundary conditions should be explicitly specified via the "
+                "'pbc' keyword argument. Defaulting to '[True, True, True]', but this will raise in v3.0.0.",
```
Comment (danielhollas) [inline src/aiida/orm/nodes/data/array/trajectory.py]: Can you use `aiida.common.warnings.warn_deprecation`?

File: src/aiida/orm/nodes/data/array/trajectory.py [hunk 4]
```diff
@@ -264,6 +294,11 @@ def symbols(self) -> List[str]:
         """
         return self.base.attributes.get('symbols')
 
+    @property
+    def pbc(self) -> list[bool]:
+        """Return the list of periodic boundary conditions."""
```
Comment (danielhollas) [inline src/aiida/orm/nodes/data/array/trajectory.py]: Could be nice to expand a bit (I copied this snippet from above)

```suggestion
        """Return the list of periodic boundary conditions.
        
        Returns a list of
            length three with booleans indicating if the structure is periodic in that
            direction. """
```

btw: why not tuple instead of a list?

**Review by danielhollas — DISMISSED** [review]
@mbercx just a quick note that I haven't forgotten about this, and plan to test this as part of the coding week. :-) 

Had a quick look and the code looks great, just some minor suggestions. 

Also, do we need to update documentation somewhere? (I haven't checked)

I haven't gone through the tests yet.

File: src/aiida/orm/nodes/data/array/trajectory.py (hunk 4 above)
Comment (mbercx) [reply src/aiida/orm/nodes/data/array/trajectory.py]: Updated! I think I just took a list to be in line with `symbols`, but `StructureData.pbc` also returns a tuple, so makes sense to be in line with that.

**Comment by mbercx** [comment]
Thanks @danielhollas! Round 3 done, let me know if there is anything else.

> Also, do we need to update documentation somewhere? (I haven't checked)

`TrajectoryData` is rather [succinctly documented](https://aiida.readthedocs.io/projects/aiida-core/en/stable/topics/data_types.html#trajectorydata) I'm afraid, but I do not intend to fix that here.

File: src/aiida/orm/nodes/data/array/trajectory.py [hunk 5]
```diff
@@ -384,7 +422,7 @@ def get_step_structure(self, index, custom_kinds=None):
                     'passed {}, but the symbols are {}'.format(sorted(kind_names), sorted(symbols))
                 )
 
-        struc = StructureData(cell=cell)
+        struc = StructureData(cell=cell, pbc=self.pbc)
```
Comment (danielhollas) [inline src/aiida/orm/nodes/data/array/trajectory.py:439]: Maybe we need something like this:


```suggestion
        try:
            struc = StructureData(cell=cell, pbc=self.pbc)
        except AttributeError:
            struc = StructureData(cell=cell)
```

**Review by danielhollas — CHANGES_REQUESTED** [review]
Really sorry for stalling this for so long. Finally got a chance to test this in AiiDAlab app and I did run into an issue. When I load a pre-existing TrajectoryData from DB, calling `get_step_structure` will raise an attribute error.

```python

~/.local/lib/python3.9/site-packages/aiidalab_widgets_base/utils/__init__.py in get_formula(data_node)
    116         # so this approach is sound.
    117         stepid = data_node.get_stepids()[0]
--> 118         return data_node.get_step_structure(stepid).get_formula()
    119     elif isinstance(data_node, StructureData):
    120         return data_node.get_formula()

~/aiida-core/src/aiida/orm/nodes/data/array/trajectory.py in get_step_structure(self, index, custom_kinds)
    423                 )
    424 
--> 425         struc = StructureData(cell=cell, pbc=self.pbc)
    426         if custom_kinds is not None:
    427             for _k in custom_kinds:

~/aiida-core/src/aiida/orm/nodes/node.py in __getattr__(self, name)
    839             return getattr(self.base.links, new_name)
    840 
--> 841         raise AttributeError(name)

AttributeError: pbc
```

I am unsure how this should be handled. How is adding a new attribute handled generally?? Is there some DB migration needed? Or just a code adjustment to account for this attribute possibly missing?

**Comment by GeigerJ2** [comment]
Hi guys, just wondering what's the status here? Is this feasible soon, or should I move it out of v2.8 and into v2.9? Thanks!

File: src/aiida/orm/nodes/data/array/trajectory.py (hunk 5 above)
Comment (mbercx) [reply src/aiida/orm/nodes/data/array/trajectory.py:439]: Thanks for testing @danielhollas! Good thing we waited, because indeed for existing `TrajectoryData` would now break. 😅 

In the past we'd probably do a database migration, but frankly they can be a pain for both users and developers. I think your proposed solution looks fine at first glance, will have another close look to see if there is anything else I missed with existing `TrajectoryData` in mind.

Comment (mbercx) [reply src/aiida/orm/nodes/data/array/trajectory.py:439]: So, with your suggested fix, just calling `.pbc` would still fail for old `TrajectoryData`. I simply adapted the `.pbc` property in  https://github.com/aiidateam/aiida-core/pull/7079/commits/21a29a9b23e4788df266c51593e03fffc4071bb7 to return `None` in case the attribute is missing (as would be the case for old `TrajectoryData`). I gave it a quick manual test, and also added some small tests in [f87afd5](https://github.com/aiidateam/aiida-core/pull/7079/commits/f87afd5344783f059ee701d7c7aa54b9eac284f7), where I emulate old `TrajectoryData` by deleting the `pbc` attribute.

Comment (danielhollas) [reply src/aiida/orm/nodes/data/array/trajectory.py:439]: Thanks, this fix makes sense. I guess the only remaining concern of not doing the migration is if some code tried to access the pbc attribute not via property but as `trajectory.base.attributes.get('pbc')`. This would still fail for older nodes right?

Comment (mbercx) [reply src/aiida/orm/nodes/data/array/trajectory.py:439]: True, but it would fail for getting any non-existent attribute. I don't see a fix for this without a migration, and I definitely don't think doing a migration for this is a good idea.

Comment (danielhollas) [reply src/aiida/orm/nodes/data/array/trajectory.py:439]: Fair, but I don't understand why migration is a bad idea? I am not saying it's a good idea, I just don't know. :-) It would be nice to see if there was a precedence for this in the past.

Comment (mbercx) [reply src/aiida/orm/nodes/data/array/trajectory.py:439]: A migration is always a pain for users, since they have to run it after simply updating their `aiida-core`, and it can take a lot of time. Also: what if they do the migration and then realise there is an issue why they have to downgrade? Not possible anymore, we have no backwards migrations. In general, we want to avoid database migrations if possible.

In this specific case, what would we set in the migration? `[True, True, True]`? That's not necessarily correct, we'd be fabricating explicit information. We could set it to `None`, but then we are really only doing a migration because `trajectory.base.attributes.get('pbc')` would fail. But that can be easily guarded agains with `trajectory.base.attributes.get('pbc', None)`. And why would a user that doesn't understand this go look for the attribute here instead of using the class attribute `.pbc`? It's right there and tab-completable. ^^

Last but not least, I really don't want to invest time into figuring out how to do migrations again. 😂  Especially because I don't think the fact that `trajectory.base.attributes.get('pbc')` failing for old nodes is a massive problem.

File: src/aiida/orm/nodes/data/array/trajectory.py [hunk 6]
```diff
@@ -131,14 +140,28 @@ def set_trajectory(self, symbols, positions, stepids=None, cells=None, times=Non
...
+            direction. The same periodic boundary conditions are set for each step.
 
         .. todo :: Choose suitable units for velocities
         """
         import numpy
 
-        self._internal_validate(stepids, cells, symbols, positions, times, velocities)
-        # set symbols as attribute for easier querying
+        if cells is None:
+            pbc = pbc or [False, False, False]
```
Comment (danielhollas) [inline src/aiida/orm/nodes/data/array/trajectory.py]: We're converting to tuple below so might as well use tuple here


```suggestion
            pbc = pbc or (False, False, False)
```

File: src/aiida/orm/nodes/data/array/trajectory.py [hunk 7]
```diff
@@ -131,14 +140,28 @@ def set_trajectory(self, symbols, positions, stepids=None, cells=None, times=Non
...
+        elif pbc is None:
+            warn_deprecation(
+                "When 'cells' is not None, the periodic boundary conditions should be explicitly specified via"
+                "the 'pbc' keyword argument. Defaulting to '[True, True, True]', but this will raise in v3.0.0.",
+                version=3,
+            )
+            pbc = [True, True, True]
```
Comment (danielhollas) [inline src/aiida/orm/nodes/data/array/trajectory.py]:
```suggestion
            pbc = (True, True, True)
```

File: src/aiida/orm/nodes/data/array/trajectory.py [hunk 8]
```diff
@@ -264,6 +293,15 @@ def symbols(self) -> List[str]:
         """
         return self.base.attributes.get('symbols')
 
+    @property
+    def pbc(self) -> tuple[bool]:
```
Comment (danielhollas) [inline src/aiida/orm/nodes/data/array/trajectory.py]:
```suggestion
    def pbc(self) -> tuple[bool, bool, bool] | None:
```

File: src/aiida/orm/nodes/data/array/trajectory.py [hunk 9]
```diff
@@ -264,6 +293,15 @@ def symbols(self) -> List[str]:
...
+        """Return the tuple of periodic boundary conditions.
+
+        Returns a tuple of length three with booleans indicating if the structure is
```
Comment (danielhollas) [inline src/aiida/orm/nodes/data/array/trajectory.py:309]: The docstring should be updated to mention the None value (and the fact that can be returned if the Node was created with earlier AiiDA version)

File: src/aiida/orm/nodes/data/array/trajectory.py [hunk 10]
```diff
@@ -82,8 +85,14 @@ def _internal_validate(self, stepids, cells, symbols, positions, times, velociti
...
                 )
-
-    def set_trajectory(self, symbols, positions, stepids=None, cells=None, times=None, velocities=None):
+        if not (isinstance(pbc, (list, tuple)) and len(pbc) == 3 and all(isinstance(val, bool) for val in pbc)):
+            raise ValueError('`pbc` must be a list/tuple of length three with boolean values.')
+        if cells is None and list(pbc) != [False, False, False]:
+            raise ValueError('Periodic boundary conditions are only possible when a cell is defined.')
+
+    def set_trajectory(
+        self, symbols, positions, stepids=None, cells=None, times=None, velocities=None, pbc: None | list | tuple = None
```
Comment (danielhollas) [inline src/aiida/orm/nodes/data/array/trajectory.py]:
```suggestion
        self, symbols, positions, stepids=None, cells=None, times=None, velocities=None, pbc: None | list[bool] | tuple[bool, bool, bool] = None
```

File: src/aiida/orm/nodes/data/array/trajectory.py [hunk 11]
```diff
@@ -82,8 +85,14 @@ def _internal_validate(self, stepids, cells, symbols, positions, times, velociti
                     'have shape (s,n,3), '
                     'with s=number of steps and n=number of symbols'
                 )
-
-    def set_trajectory(self, symbols, positions, stepids=None, cells=None, times=None, velocities=None):
+        if not (isinstance(pbc, (list, tuple)) and len(pbc) == 3 and all(isinstance(val, bool) for val in pbc)):
+            raise ValueError('`pbc` must be a list/tuple of length three with boolean values.')
+        if cells is None and list(pbc) != [False, False, False]:
```
Comment (danielhollas) [inline src/aiida/orm/nodes/data/array/trajectory.py]: This can be simplified since we've already checked the length and booleannes

`any(pbc)` will only be False if pbc == (False, False, False), therefore

```suggestion
        if cells is None and not any(pbc):
```

File: tests/orm/nodes/data/test_trajectory.py [hunk 12]
```diff
@@ -123,6 +127,8 @@ def test_trajectory_get_step_data_empty(self, trajectory_data):
         assert np.array_equal(symbols, trajectory_data['symbols'])
         assert np.array_equal(positions, trajectory_data['positions'][3, :, :])
         assert velocities is None
+        # In case the cell is not defined, there should be no periodic boundary conditions
+        assert np.array_equal(trajectory.pbc, [False, False, False])
```
Comment (danielhollas) [inline tests/orm/nodes/data/test_trajectory.py]: Simpler?
```suggestion
        assert trajectory.pbc == (False, False, False), "Trajectory without a cell should not have PBC set"
```

**Review by danielhollas — APPROVED** [review]
Okay, I've retested the current version and verified (in my AiiDAlab app) that the code works for TrajectoryData without pbc attribute. Thanks for the fix @mbercx 👍 

I am approving since this all looks great, I just have some final minor suggestions and typing fixes.

@mbercx since this file is currently not type-checked via pre-commit, can you run mypy manually?

```console
$ mypy --pretty src/aiida/orm/nodes/data/array/trajectory.py
```

You'll likely get a bunch of errors, but the number of errors should not be greater than on main. :-) 

Thanks for sticking with this and sorry for being slow to review.

Oh, it would be great to document this (not sure if TrajectoryData has a dedicated documentation?) and perhaps write down a CHANGELOG entry to help @GeigerJ2 with a release. :-)

File: src/aiida/orm/nodes/data/array/trajectory.py (hunk 9 above)
Comment (mbercx) [reply src/aiida/orm/nodes/data/array/trajectory.py:309]: Agreed! Was playing around with the various directives, and how they render (at least in VSCode). `versionadded` seems like a natural choice, but I don't really like how it looks. Any better suggestions? I can also just add it to the body, or `returns`, but I wanted to highlight this information more.

File: tests/orm/nodes/data/test_trajectory.py (hunk 12 above)
Comment (mbercx) [reply tests/orm/nodes/data/test_trajectory.py]: Hehe, sure. I just copied the line and adapted.

File: src/aiida/orm/nodes/data/array/trajectory.py [hunk 13]
```diff
@@ -34,6 +34,7 @@ class Model(ArrayData.Model):
         units_positions: str = MetadataField(alias='units|positions', description='Unit of positions')
         units_times: str = MetadataField(alias='units|times', description='Unit of time')
         symbols: list[str] = MetadataField(description='List of symbols')
+        pbc: Optional[tuple[bool, bool, bool]] = MetadataField(description='Periodic boundary conditions')
```
Comment (mbercx) [inline src/aiida/orm/nodes/data/array/trajectory.py]: @edan-bainglass let me know if you have any comments on these changes! I used `Optional` as you mentioned. Do you see any issue with this being in the `if TYPE_CHECKING:` block at the start of the file?
